pmlc/runtime/runtime_registry.h

```diff
@@ -0,0 +1,82 @@
+// Copyright 2020, Intel Corporation
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "llvm/ADT/StringRef.h"
+
+#include "pmlc/runtime/runtime.h"
+
+namespace pmlc::runtime {
+
+// Factory is used to instantiate a Runtime.
+//
+// Factory is invoked on-demand (typically well after process initialization) to
+// instantiate a runtime, either as a specific request by the PlaidML caller or
+// for general device enumeration.  The registry is responsible for arranging
+// runtimes to be instantiated exactly once; Factory implementations do not need
+// to memoize their results.
+using Factory = std::function<std::shared_ptr<Runtime>()>;
```

Comment:
Do we really need a Factory, can we just register Runtimes directly? What's the point of having an extra level of indirection?

Reply:
Added a top-level comment to address this.